### DISCUSSION TITLE
For organizations that need help

### DISCUSSION THREAD

**Discussion by zkochan** [pnpm/pnpm#3787] [General]
If your company tries to migrate to pnpm but you have some blockers, add a post here and reference the issues that you have created. We will try to prioritize bug fixes/feature requests that are coming from big teams/companies.

The post should contain the name of the company and the size of the team.

Feel free to use this discussion also if you already migrated to pnpm but really need help with some issues.

**Comment by TeChn4K** [comment]
For us, the most challenging thing is being able to export/dockerise one workspace/package and its (internals AND externals) dependencies without packing the whole repository.
`make-dedicated-lockfile` is unusable because of https://github.com/pnpm/pnpm/issues/3442
`pnpm-isolate-workspace` is externally maintained and also has some (less problematic) issues.

A lot of discussions about this topic, but no mature solution yet : https://github.com/pnpm/pnpm/issues/3114 https://github.com/pnpm/pnpm/issues/1637 https://github.com/pnpm/pnpm/issues/2198

_small company / few IT members_

**Reply by zkochan** [reply]
Track this issue: https://github.com/pnpm/pnpm/issues/4378

**Comment by EscapeB** [comment]
For us, we are using pnpm to manage dependencies for our big monorepo (200+projects and 100+developers).

The problems we are facing are: 

**1. Workspace single lock file**
Cause we are using pnpm workspace, single lock file was generated by default. After migrate lots of projects into monorepo, the lockfile became unstable and always get git conflict because there are many developers doing their work at same time.

We plan to turn off this option [shared-workspace-lockfile](https://pnpm.io/workspaces#shared-workspace-lockfile), but still have some blockers. I have already list them.

https://github.com/pnpm/pnpm/discussions/4247


**2. Lockfile update**

Lockfile not updated after run `pnpm install` if we change pnpm option or modify pnpfile.cjs. The recommend way is to regenerate 
a new lockfile, but we want to keep the lock version for packages.


Bytedance / (100+ developers)

**Reply by zkochan** [reply]
> 2. Lockfile update

We can fix this. However, what do you need the pnpmfile for? We have lot's of other ways not to change the dependencies: overrides, packageExtensions, peerDependencyRules.

**Reply by EscapeB** [reply]
Really appreciate for your support !

We need use pnpmfile to do some resolution work and fix flaw package which has some phantom or incorrect dependencies. Some of them needs to be calculated dynamically so we can't put them in package.json.

**Comment by erango** [comment]
Hi! We're having trouble with unlinking. Our use case: we have a ton of packages in a monorepo, and they have some interdependencies. Sometimes two levels of dependency. To work more efficiently, if we don't need local code from the dependencies, we use repo versions instead. That way, you don't have to install the entire dependency tree just to work on one package.

We have two unlink-related bugs that create havoc on a daily basis for us as there are 70+ developers making changes in that repo every day.

Thanks so much for your attention! ❤️ 

https://github.com/pnpm/pnpm/issues/4294
https://github.com/pnpm/pnpm/issues/4295

**Comment by tgroutars** [comment]
Hi! 

We are trying to migrate our monorepo to `pnpm` from `yarn` + `lerna`, but we're stuck `import`ing because pnpm doesn't respect the versions specified in our nested `yarn.lock`s. 

Explanation + minimal repro in #4385 

Thanks a lot in advance 🙏 

Maze / 35+ engineers

**Reply by tgroutars** [reply]
I managed to tweak the code to make it work for my usecase, although my code is not memory efficient, very ugly and definitely breaks other parts of the package.

I'm not opening a PR on this repo for that reason, but you can have a look at the one I created on my fork if you want some inspiration https://github.com/tgroutars/pnpm/pull/1

Thanks a lot for what you do for pnpm, and I hope you're safe 🙏

**Reply by EscapeB** [reply]
I successfully used `.pnpmfile.cjs` to lock the dependencies version in migration process of our monorepo, may helps you without changing pnpm source code.

Here are my steps:

1. **Parse previous lock file generated by  (yarn / npm / pnpm)**. You can use some useful libs in community to parse different kinds of lock file. I used [this](https://github.com/snyk/nodejs-lockfile-parser) 
2. **Generate a temp `.pnpmfile.cjs` in monorepo root (or in each project's folder if you are using separated lock file)**. In `.pnpmfile.cjs`, we can do resolution logic in [readPackage hooks](https://pnpm.io/pnpmfile#hooksreadpackagepkg-context-pkg--promisepkg) based on lock version data collected from step1.
```javascript
module.exports = {
  hooks: {
    readPackage,
  },
};
// data struct injected in step1
const resolutions = {
  pkg1: {
    version: '1.0.0',
    dependencies: {
      dep1: '1.0.0',
      dep2: '1.0.0',
    },
    devDependencies: {
      dep3: '1.0.0',
      dep4: '1.0.0',
    },
  },
};

function readPackage(packageJson, context) {
  //  do your specific logic to modify packageJson based on const "resolutions"
  return packageJson;
}

```
3. Run pnpm install to see if it works.

**Reply by tgroutars** [reply]
Thanks! I managed to make `pnpm import` work for my usecase, though. I need to do it once, so I'm good now!

**Comment by yuyez** [comment]
~We're migrating from `yarn` to `pnpm` and ran into a possible bug #4592. Appreciate any pointers!~

We have migrated from yarn to pnpm!

We're a team of ~60 devs in Microsoft.

**Reply by zkochan** [reply]
Great! I am happy it worked for you!

**Comment by raygesualdo** [comment]
We've migrated to env vars for private registry credentials and when a teammate doesn't have the env var set, `pnpm` is exhibiting unexpected behavior: #4806.

We're a team of ~150 devs at @Salesloft.

**Comment by joscalv** [comment]
Hi,

We are migrating to pnpm in Siemens Gamesa in a team of 100 dev aprox, but we found this issue:

We have a library with common utilities to develop and with many peer-dependencies to be installed by every teams.
The option "auto-install-peers=true" is very convenient for us, because the peer dependencies of that library are installed automatically without doing anything by every team.

The problem is that pnpm only installs that peer dependencies when pnpm-lock.yaml don't exists. If we delete node_modules after or in the CI/CD system, that dependencies are not installed.

Thanks for your help.

**Comment by sbonasu** [comment]
Hi,
   We are actively working on migrating big monorepo with 200+ packages to pnpm from yarn@1.x but ran in to a blocker(Issue # [4297](https://github.com/pnpm/pnpm/issues/4927). 

large company/200+ team

**Comment by jerrychen4git** [comment]
hi,
we have cyclic workspace dependencies but don't show any detail info about cycle, it's hard to check cycle manually , please fix this issue
https://github.com/pnpm/pnpm/issues/5059

bytedance/100+ team

**Comment by apaniel** [comment]
hi, 
we have a monorepo in a company with 30+ developers working on it at the same time, we are having some unexpected updates in the lockfile that as side effect are producing not needed redeployment of different containers

https://github.com/pnpm/pnpm/issues/5108

**Comment by stefan-toubia** [comment]
Inconsistent/buggy behavior when `shared-workspace-lockfile` is disabled has been the main blocker migrating our repo. Shared lock file is not an option, monorepo contains multiple independent projects. From what I've seen it's not a hard blocker per se, biggest issue we're seeing so far has popped up in multiple issues like #4245 and #4868. We can likely work around this issue for some time since it just causes install times to be slower.

The main concern right now is uncertainty about how well `shared-workspace-lockfile=false` will be supported going forward. The docs mention nothing about this, but I've seen a number of comments recommending people not use this setting and that its buggy/not well tested. It would ease our concerns a bit if the pnpm team can comment on whether theres any plan to address these issues.

I should really also mention that from what we've seen so far, pnpm is a huge improvement over yarn and we're seeing a huge speed up in local and CI install times, so big shout out to the pnpm team.

Details on our monorepo:
100+ engineers, multiple independent teams, mixed langauges. Just a subset of those teams are working with Node. Currently using `yarn` v1.

**Reply by zkochan** [reply]
> The main concern right now is uncertainty about how well shared-workspace-lockfile=false will be supported going forward. 

I don't have an answer about it now. I don't plan to remove it but I also don't have anything related to it in priority. Maintaining this many settings is hard and ideally I want to reduce the amount of variations when possible.

**Reply by ProTip** [reply]
I would like to echo these sentiments. Biggest issue I'm running into is brining packages into a workspace that have patches and override rules that should only apply to that package.

If the `overrides`, `patchedDependencies`, and `packageExtensions` could be scoped to individual workspaces `packages` it would solve a lot of headaches for me.

**Reply by giancarlosisasi** [reply]
omg, I didn't know that overrides don't apply to individual workspace packages 😱 . I'm going to review my package.jsons right now 🏃🏼‍♂️

**Reply by ProTip** [reply]
It appears `overrides` can be used to override a specific dependency at the top level of a workspace package.

They can also be used to drill into a workspace package's dependencies and override a single occurrence there.

However, it can't be used to override every occurrence of a dependency recursively within a workspace package. This turns into a huge fiasco when you want to override a dependency that's part of a suite of packages from a project which gets depended on numerous times. You basically have to locate every package that requires it and add a separate `overrides` entry :|

**Comment by JamesSeymour** [comment]
We are attempting to migrate from yarn to pnpm. We have a monorepo with 200+ packages, but we are seeing an issue when some of the packages have different versions of the same dependency. - https://github.com/pnpm/pnpm/issues/5176

We would appreciate any help in understanding the issue more, so we can migrate our codebase!

Caplin/50+ team

**Reply by JamesSeymour** [reply]
Just thought I would update this. We moved our full codebase over to PNPM in September after a few minor workarounds and fixes on our side. It has been flawless ever since, our developers love working with it, and it really has increased our efficiency working on a large monorepo. Great work!

**Comment by lefreet** [comment]
is there a plan to auto version like `lerna verion` ？ 

we have some package depend on each other.  for our project with `workspace`，the pnpm provide a link like `"foo": "workspace:*"`  so that they all right when runtime.   but when we want to publish them, `pnpm publish`, they published immediate with my `admin role`  in out pravite registry.  so, packages be overwrited :(   .    

so,  there area some way to increase the packages's version  by git tag check like `lerna`.    with the `vue | vite` resolve these with custom  `scripts/release.js` .  -> https://github.com/vuejs/core/blob/main/scripts/release.js

**Reply by zkochan** [reply]
No, you should use [changesets](https://pnpm.io/using-changesets) or [rush](https://rushjs.io/) for versioning.

**Comment by d4mr** [comment]
We are using using Turborepo with pnpm workspaces. We have already migrated over to pnpm, but are facing issues with package duplication.

Detailed issue: https://github.com/pnpm/pnpm/issues/5351
Company: Realto
Team Size: 15

**Comment by soundstep** [comment]
We are building a new CI tool, we are going to use the npm publish command until the pnpm one is fixed.

Issue: https://github.com/pnpm/pnpm/issues/5396
Company: ITV UK
Team Size: 100+

**Comment by tomdavidson** [comment]
Trying to adopt pnpm but failing on multiple fronts.

- Workspace filtering: https://github.com/pnpm/pnpm/issues/5438
- Peer Dependencies: https://github.com/pnpm/pnpm/discussions/5366

Company: RSQ
Team Size: 5

**Comment by giancarlosisasi** [comment]
Migrating from yarn and lerna we found this error:

-  #5479

Company: crehana
Team size: 50+

**Comment by walkerburgin** [comment]
While trying to upgrade a large monorepo from pnpm 6.x to 7.x we hit this issue: 
+ https://github.com/pnpm/pnpm/issues/5493

Company: Palantir
Team size: Several hundred

**Comment by MilesWellsSVT** [comment]
Not so much a blocker as an annoyance:

https://github.com/pnpm/pnpm/issues/5507

Company: SVT Robotics
Team size: ~20

**Comment by sbonasu** [comment]
This is a blocker

#5610

Company: Cyber Security
Team Size: 150-200

**Comment by mikeymop** [comment]
Hello,
We use a self hosted artifactory instance for our dependencies.
When using pnpm `7.6.0` pnpm is able to authenticate fine. However, after upgrading to the latest pnpm I started getting authentication errors.

We currently use pnpm for our monorepos and for ci on many of our projects. I've notified the teams not to update past `7.6.0` for now.

```
$ pnpm --version
7.14.2
$ pnpm add -D react-test-renderer
Packages: +1117
++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
 ERR_PNPM_FETCH_403  GET https://artifactory.example.com:443/artifactory/api/npm/javascript/react-test-renderer/-/react-test-renderer-18.2.0.tgz:  - 403

No authorization header was set for the request.

These authorization settings were found:
//artifactory.example.com/artifactory/api/npm/javascript/:_authToken=eyJ2[hidden]
```
```
$ pnpm add -g pnpm@7.6.0
$ pnpm add -D react-test-renderer
Packages: +1117
++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
Packages are hard linked from the content-addressable store to the virtual store.
  Content-addressable store is at: /Users/$USER/Library/pnpm/store/v3
  Virtual store is at:             node_modules/.pnpm
Progress: resolved 2254, reused 1112, downloaded 3, added 3, done

devDependencies:
+ react-test-renderer 18.2.0
```

**Reply by mikeymop** [reply]
Hello just an update.
This issue seems to be resolved with `pnpm@8.0.0`.

Thank you for your continued work on this project!

**Comment by sgoldgar** [comment]
This has been a blocker for us adding new packages to our monorepo and we do have workarounds but they are not ideal

Issue: @types/react versions mismatch with upstream transient dependencies 
[https://github.com/pnpm/pnpm/issues/5736](https://github.com/pnpm/pnpm/issues/5736)

Company: Disney Streaming
Team size: ~150

**Comment by dominikg** [comment]
this issue affects vite-ecosystem-ci 

https://github.com/pnpm/pnpm/issues/5754

**Comment by swernerx** [comment]
This issue affects our internal development pipeline at EnBW.

https://github.com/pnpm/pnpm/issues/5789

**Comment by sgoldgar** [comment]
Hello, we have another issue (non-blocking this time) for Disney Streaming that we'd love some advice on!

**Issue**: Changes when running pnpm install not reflected when running pnpm install --frozen-lockfile

[#5794 ](https://github.com/pnpm/pnpm/issues/5794)

Company: Disney Streaming
Team size: ~150

**Comment by minonl** [comment]
Hi, we are caching store locally on CI runner, this issue blocks us from upgrade pnpm version.

https://github.com/pnpm/pnpm/issues/5797

Company: Leyantech
Team size: ~150

**Comment by 35ujq435jq45** [comment]
Hi there, we want to verify our lockfile in CI. However, pnpm does not offer such a feature. Hence, we tried `pnpm install --frozen-lockfile`. But, this will download and install all dependencies which we want to avoid.
So we tried `pnpm install --frozen-lockfile --lockfile-only`. However, this will modify the lockfile :-S

Company: NVIDIA
Team size: ~70

Bug report: https://github.com/pnpm/pnpm/issues/6094

**Comment by vajahath** [comment]
A fresh [Angular](https://angular.io/) project with [AngularFire](https://github.com/angular/angularfire) using pnpm does not build due to incorrect nested dependency resolution. But it works with npm.

This seem to have an impact on Angular community as a whole (who are trying to use pnpm instead of npm).

Bug report: #6165

**Comment by trappar** [comment]
Company: Curology
Team size: ~50 devs

We recently started using PNPM in a new monorepo and we are pretty hard blocked by these two issues:
* https://github.com/pnpm/pnpm/issues/6221
* https://github.com/pnpm/pnpm/issues/1801

Edit: We are unblocked.

**Comment by millsp** [comment]
Hey Zoltan :wave:

We (Prisma) would appreciate your help on this issue
* https://github.com/pnpm/pnpm/issues/6272

Thank you for your time!

**Comment by jakebailey** [comment]
Kinda ridiculous but I'm working on converting all of DefinitelyTyped to a big monorepo, but I'm hitting some perf troubles.

- #6277

**Comment by supersabillon** [comment]
Hello, we would like to use the dedupe-peer-dependents=true but running into this issue: https://github.com/pnpm/pnpm/issues/6295

Company name: Disney Streaming
Team size: ~100 devs

**Comment by chemdrew** [comment]
Hi, working on a large monorepo and running into differences with the file structure of the installed modules when using pnpm vs npm that is throwing errors when importing them. Workaround for now is using v6 of pnpm.

Company: Cameo
Team size: ~50

Bug report: https://github.com/pnpm/pnpm/issues/6378

**Reply by zkochan** [reply]
Ha, I have just used Cameo today for a birthday message😆

[a comment by an automated account is not shown]

**Comment by josh-hemphill** [comment]
We have devs doing their work on a multi-user VM with all the working directories for projects under a windows share mounted as a drive. So to facilitate multiple users simultaneously accessing the global store and cache etc. we need to run the store server. It runs fine when pointed to the global install directory on the standard C: drive, but since almost all the calls to it are coming from a mounted drive, it needs to be using the store on that drive to prevent unnecessary duplication.

Organization: US Fed Gov
Team size: 3 teams of 3-5 devs (2 more teams might move over to pnpm later) ~11

Bug report: #6386

**Comment by jakebailey** [comment]
Awkward node_modules hoisting/etc problem for DefinitelyTyped: https://github.com/pnpm/pnpm/issues/6457

**Comment by SeinopSys** [comment]
`@pnpm/*` dependencies are gone from NPM breaking all of our rush monorepo builds: https://github.com/pnpm/pnpm/issues/6545

**Comment by 123joshuawu** [comment]
We are using `pnpm` for our monorepo which is packaged and deployed in customer environments. https://github.com/pnpm/pnpm/issues/6584 is preventing us from creating self-contained packages.

Company: Aviatrix Systems
Team Size: ~20 devs

**Comment by Macariel** [comment]
Seems like some breaking changes have been introduced with newer pnpm versions (8.6.1): #6648 
Easily fixed with an update for an individual developer, but eats at our time to debug the problem locally and might pose big problems with our pipelines

Team Size: ~30 devs

**Reply by zkochan** [reply]
Yes, there was an error in the code that checked the lockfile version. Unfortunately I discovered the issue only after releasing the new lockfile format version. This shouldn't happen again. You will all have to update to v8.6. Sorry for the inconvenience. It is too late to ship a fix to v8.5 as v8.6 is already latest.

**Reply by ckolumbus** [reply]
Hi everyone,

we (Development department within Siemens Healthineers) are migrating our frontend development from npm to pnpm. 

TL;DR: we need #6679 , currently 10 devs impacted (but we have a workaround) but 150 devs if we would go for full roll-out

Details:
Due to "controlled build" requirements we do not normally use public npm registries but isolated internal hosted feeds (with no upstream connection to npmjs or any other repo). for this to work we have the registry configure as being read from a variable, and because our feeds are all authenticated, we also need to configure the auth token. With this we can change the feed based on development phase or release stage (QA feed, Product feed, dev feed) without changing the .npmrc file in the git repo. 

example:
```
registry=https://${NPM_REGISTRY}/
//${NPM_REGISTRY}/:_authToken=${NPM_TOKEN}
```

The problem:
`npm` actually replaces both `NPM_REGISTRY` references, `pnpm` replaces only rhs variables, so the one for the token config remains unreplaced. 

Thanks for this cool piece of software, it really helps to improve and speed up our development. we can also provide a PR if this helps.

**Comment by icy0307** [comment]
Hi, we are using `pnpm` for a large web application with lots of developers come and leave.
Being able to modify `package.json` via hook function instead static field is an awesome feature, as we can prevent invalid package when installing.
For example:
1. package with no author  and repository field.
2. package has wrong or fixed prod dep and peer dep.

However, both lint and  changeing package dependencies to future release compatible version requires other package's help
like `semver`.
`.pnpmfile.cjs` currently don't support this as it would cause a loop which require package before read package .

See  #6731 for more details.

Team Size: ~100 devs
Company: Tencent

**Comment by kevinpastor** [comment]
Hello,

We are using PNPM in order to manage one of our monorepo. In order to test the integration of our libraries in other internal projects, we would like to make use of `pnpm link`.

We have a problem with linking and peer dependencies. Peer dependencies behaves likes dependencies. Our specific use case is with React as a peer dependency.

This problem not only affects my team, but our whole department who makes use of our ecosystem of libraries.

See https://github.com/pnpm/pnpm/issues/6804 for more details.

Company: Ubisoft
Team size: ~250 developers

**Reply by KSXGitHub** [reply]
I think the proper solution would be to publish your packages to a private registry. Is there any reason this isn't sufficient for your use case?

If I change https://github.com/kevinpastor/pnpm-link-peers-issue to using `pnpm-workspace.yaml` and add `some-library` to the dependency list of `some-consumer/package.json`, `pnpm start` works properly (prints "hello world").

If you can combine all your related internal projects in one workspace, it can be solved.

**Reply by kevinpastor** [reply]
I think your solution seems simply to be a workaround for an underlying issue of `pnpm link`. We would expect peer dependencies to behave the same as they do everywhere else with PNPM.

To give you a bit more context, in our actual solution, we have a monorepo (a project with a `pnpm-workspace.yaml`) for all of our common libraries and then a single monorepo per microservices. We ended up with this architecture in order to simplify our release process and to limit visibility on some parts of the codebase.

zkochan suggested us in the issue I created to use the `file:` protocol, which seems to better fit our needs.

**Comment by benjibenoit** [comment]
Hello, we would like to use pnpm with multicontainer docker setup: https://github.com/orgs/pnpm/discussions/6812

Company name: Natiq
Team size: 3 devs

**Comment by antony** [comment]
Hello,

we're encountering trouble telling pnpm to use a previously downloaded and cached version of nodejs on github actions, which means that builds frequently (10% of the time?) stall and time-out, because the nodejs site seems to be unreliable. It's costing us a great deal of wasted time.

Company name: Beyonk
Team size: 7 devs

Issue raised here: https://github.com/pnpm/pnpm/issues/6926

**Reply by KSXGitHub** [reply]
https://github.com/pnpm/pnpm/issues/6926#issuecomment-1671157032

**Comment by savolkov** [comment]
Hey!

We faced issues with dependencies resolution after migrating from v7 to v8

Company: GreenData
Team size: 25 FE Devs

Issue #6944 

Thanks for your work <3

**Comment by jakebailey** [comment]
Goofy bug for DefinitelyTyped here: https://github.com/pnpm/pnpm/issues/7283

This affects DT's use of partial `pnpm install`s, as some needed deps in the workspace are not installed.

**Comment by jfirebaugh** [comment]
`pnpm import` does not preserve locked versions from package-lock.json: https://github.com/pnpm/pnpm/issues/7287

This is making a migration from npm to pnpm considerably more difficult.

Company name: Figma
Team size: ~400

**Comment by andschub-ms** [comment]
Absolute paths in pnpm-lock.yml when using resolutions #7858

Company: Microsoft
Impact: We are trying to onboard to pnpm at an org-level, and this issue is causing us difficulties.
https://github.com/pnpm/pnpm/issues/7858

**Comment by gcampbell-msft** [comment]
Issues with upgrading to pnpm version 8.15.7, where `tar` is still being installed as `6.2.0` instead of `6.2.1`. #7939 

Company: Microsoft
Impact: Trying to update to `tar` version 6.2.1

**Comment by andschub-ms** [comment]
"pnpm.exe update @myscope/mypackage --latest --recursive" is updating all packages instead of mypackage #8329

Company: Microsoft
Impact: We were using this command in automation and it is now broken after moving from pnpm v8 to v9. Behavior of pnpm v9 update command is also now inconsistent with the public documentation. Would appreciate taking a look or helping to provide a quick work-around. Thank you.

**Comment by patricktree** [comment]
"Make `shared-workspace-lockfile=false` well supported and as fast as `shared-workspace-lockfile=true`" https://github.com/pnpm/pnpm/issues/8334

Company: hokify
Impact: Greatly improve development velocity

Thanks!

**Comment by chenyulun** [comment]
onyl install `vue-loader@15.11.1`, `pnpm-locak.yaml` has `@vue/compiler-sfc`???

`node_modules/vue-loader/package.json` no have `peerDependencies['@vue/compiler-sfc']`, but `pnpm-locak.yaml` has `@vue/compiler-sfc`
<img width="1497" alt="image" src="https://github.com/user-attachments/assets/2b3f5e67-1c6c-424d-8896-eca5c2f0d657">

**Reply by chenyulun** [reply]
why use `@yarnpkg/extensions`???
`vue-loader@15.11.1` support `@vue/compliler-sfc@^2.7.0`

**Reply by zkochan** [reply]
If you don't want to use `@yarnpkg/extension`, disable it with [ignore-compatibility-db=true](https://pnpm.io/npmrc#ignore-compatibility-db)

**Reply by chenyulun** [reply]
> If you don't want to use `@yarnpkg/extension`, disable it with [ignore-compatibility-db=true](https://pnpm.io/npmrc#ignore-compatibility-db)

@zkochan hi, but I don't want to set [ignore-compatibility-db=true](https://pnpm.io/npmrc#ignore-compatibility-db),
use like this:
```json
"pnpm": {
    "packageExtensions": {
      "vue-loader": {
        "peerDependencies": {
          "@vue/compiler-sfc": "^2.7.16"
        }
      }
    }
  }
```
but it didn't working ,  `@yarnpkg/extension` is too high a priority
```js
      [`vue-loader@<=16.3.3`, {
        peerDependencies: {
          "@vue/compiler-sfc": `^3.0.8`,
          webpack: `^4.1.0 || ^5.0.0-0`
        },
        peerDependenciesMeta: {
          "@vue/compiler-sfc": optionalPeerDep
        }
      }]
```

<img width="1470" alt="image" src="https://github.com/user-attachments/assets/3ff1d1c4-6e42-40a3-8448-c1d02f41479c">

**Reply by chenyulun** [reply]
Priority levels should be adjusted？

**Comment by andrei-picus-tink** [comment]
https://github.com/pnpm/pnpm/issues/6893 pnpm does not fail when missing peer deps with `strict-peer-dependencies` (duplicate https://github.com/pnpm/pnpm/issues/7087)

`strict-peer-dependencies` is essentially broken, leading to invalid lockfiles and crippling tools like dependabot and renovatebot.

Company: Tink, a Visa solution
Size: ~50/500

**Comment by lfloyd117** [comment]
We cannot upgrade PNPM from v8 to v9 due to `pnpm dedupe` hanging in v9. I have made an issue here: https://github.com/pnpm/pnpm/issues/8446 

Medium/large company, 200+ developers

**Comment by lfloyd117** [comment]
Feature request for an issue with PNPM's resolution that has been causing us a lot of issues as of late:
https://github.com/pnpm/pnpm/issues/8502

Medium/large company, 200+ developers

**Comment by andschub-ms** [comment]
Cannot upgrade to v9.10.0 due to error printing in the console after install
#8538 

https://github.com/pnpm/pnpm/issues/8538

**Comment by lfloyd117** [comment]
We are having an issue with `pnpm install` hanging on the latest release when changing package versions.
https://github.com/pnpm/pnpm/issues/8720 

Medium/large company, 200+ developers

**Comment by andschub-ms** [comment]
Microsoft: hundreds of devs
Ran into a pnpm-cache corruption bug, and we still can't upgrade to latest from 9.9.0 due to PNPM install crashing.
Cache Corruption
#8734 
Install crashing
#8538

**Comment by fpapado** [comment]
We have a large monorepo with `shared-workspace-lockfile=false`. We want to delegate `auditConfig` to each workspace package manifest, but currently `pnpm audit` only consults the root manifest.

Full details at: https://github.com/pnpm/pnpm/issues/8736

Large company, ~100 contributors in the relevant monorepo.

**Comment by haines** [comment]
Hi, I'm from [Cerbos](https://www.cerbos.dev). We've run into a few issues with using `pnpm deploy` to build Docker images:

- https://github.com/pnpm/pnpm/issues/5315
- https://github.com/pnpm/pnpm/issues/7683
- https://github.com/pnpm/pnpm/issues/8778

**Comment by mdjastrzebski** [comment]
Hello, I've found an issue with publishing packages from our monorepo: #8862.

This blocks us from using pnpm to publish React Native template containing executable `gradew` file.

Company: [Callstack](https://callstack.com/)
Team size: 5

**Comment by sougiovn** [comment]
I found a bug with `peerDependencies` not resolving alias: https://github.com/pnpm/pnpm/issues/9017

This block us from upgrading to v10 because our internal packages depend on aliasing peerDependencies.

Large company, hundreds of devs

**Comment by dennemark** [comment]
we are struggling with building via rollup. some packages like `@lexical/core` have dependencies like `@lexical/mark` that can be imported in a repo. when building rollup cannot resolve those imports, unless i specifiy in `package.json` that i want it injected. like this: 
```ts
"dependenciesMeta": {
    "@lexical/core": {
      "injected": true
    },
  },
```

but if i do this for a patched dependency, this does not work anymore.

**Comment by jwanner83** [comment]
We are using the standalone version of pnpm companywide because it is cross platform compatible and works great. A lot of developers who used nvm-sh before, miss the possibility to set a specific node version which is only used in the current terminal session and not globally. 

Issue: https://github.com/pnpm/pnpm/issues/9364
Discussion: https://github.com/orgs/pnpm/discussions/3434#discussioncomment-12674336

Company: clavis IT ag
Team Size: 30+

**Comment by antitoxic** [comment]
We are using `--silent` flag to pipe output of `pnpm run` and parse it, but `@pnpm/log.group` prints `groupstart` CI markers even if  `--silent` is used inside GH actions, so our logic now breaks.

Issue: https://github.com/pnpm/pnpm/issues/9563

Company: Yotpo
Team Size: 100+

**Comment by jcrben** [comment]
I work for a large company and would like to take advantage of pnpm's version management feature https://pnpm.io/cli/env

I need something like `NVM_NODEJS_ORG_MIRROR` https://github.com/nvm-sh/nvm?tab=readme-ov-file#use-a-mirror-of-node-binaries - does that exist for pnpm?

If so, it would be worth documenting on the page

**Reply by zkochan** [reply]
Yes, there is a setting for that: https://pnpm.io/settings#node-mirror

**Comment by tkolanko** [comment]
@zkochan I am looking at using pnpm but all of our scanning tools rely on a package-lock.json. Similar requirement to https://github.com/orgs/pnpm/discussions/3367#discussioncomment-4214798

is it possible to take a pnpm lock file and auto genereate a package-lock based on it

**Comment by Exelord** [comment]
Hi @zkochan! We are using pnpm v10 (10.19.0) and noticed a discrepancy between `pnpm install` and `pnpm install --frozen-lockfile`.

In the following scenario `Package 1 <- Package 2 <- tsx` the `pnpm install` will expose `tsx` binary (`node_modules/.bin/tsx`) to the Package 1, while `pnpm install --frozen-lockfile` will not. This creates a few issues where we get a "random" version of tsx if several dependencies will have it as their dependencies, and a logic that works locally but not on CI (tsx being present in consumer app locally).

My expected behavior would be the  `pnpm install --frozen-lockfile` case where sub-dependecnies bins are not exposed to consumer.

Could you please help me clarify if this is a bug and which behavior is expected?

**Comment by adamdehaven** [comment]
We've noticed an issue with `pnpm.overrides` utilizing references (with `$`) are reset to an actual version after running `pnpm audit --fix` even if the dependency is not impacted by the audit: https://github.com/pnpm/pnpm/issues/10325

**Comment by kweij** [comment]
The instruction on [how to use `pnpm` on CI/CD](https://pnpm.io/continuous-integration) are all based on first installing `corepack` using ... `npm`. We're piloting using `pnpm` for security reasons, because running `npm install [...]` has it's risks (insert recent supply chain attacks here).

Is there al alternative best practice to installing `pnpm` (consistently for a certain version) for CI/CD without using `npm`?

**Comment by ronjouch** [comment]
Dependabot is partially broken with pnpm (v10, haven’t tried earlier versions): although their [support matrix says Security updates are supported](https://docs.github.com/en/code-security/reference/supply-chain-security/supported-ecosystems-and-repositories#supported-ecosystems-and-repositories), PR creation fails with this error:

> **Dependabot doesn't support the 'updating transitive dependencies' feature for pnpm package_manager**
>
> Dependabot attempted to update your dependencies but encountered an unsupported feature: 'updating transitive dependencies' for pnpm package_manager. Currently, this feature is not supported by Dependabot.

Dependabot issue, with 25 upvotes as of writing: https://github.com/dependabot/dependabot-core/issues/13177

**Comment by sresch4b** [comment]
We're struggling with the supply-chain checks that pnpm offers in combination with Azure Artifacts, because it manipulates the meta data of packages from upstreams like npm.

#11720 

Large company
Teamsize ~20 developers, but I guess everyone who is using pnpm in combination with Azure Artifacts feels the same pain.

**Comment by ajeetdsouza** [comment]
pnpm 11.4.0+ breaks installs for the [Buf Schema Registry](https://buf.build/product/bsr). This issue affects not only our company but also several of our customers, many of whom are large enterprises.

https://github.com/pnpm/pnpm/issues/12145

**Comment by kj800x** [comment]
HubSpot is in early phases for migrating away from an in-house custom package manager to pnpm. So far we have had strong success. We're making heavy use of the config plugin and custom resolver features of pnpm 11 to help bridge between our legacy custom package manager and standard modern packages that we host on an internal verdaccio server.

Organizational context: HubSpot currently has about 2k engineers, with about ~800 engineers that do some form of JavaScript / TypeScript development. We have 4064 packages in our custom package manager. We use a micro-frontend SPA app architecture with library packages for shared logic. Our repos typically have around 1-10 packages configured together in a workspace.

Generally the migration has been going smoothly. One feature so far would make it easier for us to use our config plugin in all the contexts we need it: pnpm#12180

**Comment by RomanHotsiy** [comment]
- Issue: #11859
- Company: Redocly
- Team Size: 60+

**Comment by astegmaier** [comment]
All Microsoft engineers are currently blocked from using pnpm's self-update feature with versions of pnpm > 11.5.3. (i.e. if you have pnpm>=11.5.3 installed globally and you run `pnpm` in a repo that has `"packageManager": "pnpm@X"`, pnpm will fail to auto install version X).

This is caused by:
1. Two issues in pnpm@11.5.3:
    a. #13619 (introduced by #12296)
    b. #13147 (introduced by #12292)
2. A central security team recently started blocking public npm registry access on all developer machines at the Microsoft, and forcing all traffic to go through a feed proxy that centrally controls quarantining and auditing packages for vulnerabilities.

A reproduction and detailed explanation is available at https://github.com/astegmaier/playground-pnpm-package-proxy

This affects >200 monorepos and >4k engineers.

**Reply by zkochan** [reply]
ok, I will look into it.

**Reply by astegmaier** [reply]
Thanks so much! I tested out 11.21.0 (which contains #13728), and it works great :-)

**Comment by tw-blue** [comment]
I created a discussion on the topic of potentially insecure packages being downloaded [Here](https://github.com/orgs/pnpm/discussions/13078#discussion-10440628)
Which prevents us from migrating to pnpm right now

I work at domeba GmbH which has about 140 Employees

**Comment by ByronDWall** [comment]
Issue: https://github.com/pnpm/pnpm/issues/13715
This issue affects multiple monorepos in our organization, and roughly 200 engineers.

**Comment by jamenh** [comment]
We'd like to migrate to pnpm 12 from Yarn 4 on one huge frontend monorepo. Current blocker mainly is lockfile update being slower than our Yarn 4 baseline. Hope performance issues could be addressed, raised https://github.com/pnpm/pnpm/issues/14352.

Company: Atlassian
Devs working on the monorepo: 2500+

**Comment by lauracallahan-wk** [comment]
Issue: https://github.com/pnpm/pnpm/issues/14747
Has a workaround, but impacts our org of ~500 engineers.
